### DISCUSSION TITLE
Retaining the persisted data after failed retries

### DISCUSSION THREAD

**Discussion by mitioshi** [TanStack/query#6510] [Q&A] [answered]
I need some help with react query and the async storage provider. When a page is initially loaded, the state is correctly read from the persisted cache. If the maximum number of retries has passed due to server errors (on the screenshot), then the whole saved state in PersistedClient.clientState.queries is reset. How can I make it so that here the state does not change after all retries? btw, in the GUI the state remains correct, nothing is reset. The problem is only with persister
![image](https://github.com/TanStack/query/assets/19928439/e1ff3a70-70a3-41af-8d51-4ce9cf3cfba6)

**Comment by TkDodo** [comment]
> then the whole saved state in PersistedClient.clientState.queries is reset. 

not sure what you mean by that. Have you implemented a custom retry logic? Please show a minimal, runnable reproduction.

**Reply by mitioshi** [reply]
I haven't implemented a custom retry logic.
Here's the example https://codesandbox.io/p/devbox/laughing-neumann-nl23s9. 
To reproduce the issue:
1. first let the page load the data. The query will be stored under the REACT_QUERY_OFFLINE_CACHE key in localStorage. 
2. Then, uncomment line 22 and save the file. This will reload the page. 
3. Wait for ~5 secs and look at the `REACT_QUERY_OFFLINE_CACHE` key in localStorage. It will NOT contain the query result fetched in step 1

At step 3 I want to keep the old query in local storage. How do I achieve that?

**Reply by TkDodo** [reply, marked as answer]
queries that are in error state will not be persisted per default, because errors are not serializable. If you get a background refetch error, your query will be in the following state:

```
status: 'error'
error: YourErrorHere
data: YourStaleDataHere
```

if we were to persist that, error would probably become `null` because it can't be serialized to JSON. Then, if you restore again, you state would be:

```
status: 'error'
error: null
data: YourStaleDataHere
```

However, `status:error` is defined as "A Query with an error", so you _should_ be able to access `error.message` - but this will error at runtime in this scenario.

So, long story short, per default, you can't. You can overwrite our default persistence behaviour to also persist queries that are in error state:

```
<PersistQueryClientProvider
  client={queryClient}
  persistOptions={{
    persister,
    dehydrateOptions: {
      shouldDehydrateQuery: (query) =>
        // custom logic here, default is defaultShouldDehydrateQuery(query)
    },
  }}
>
```

The default function is implemented as:

https://github.com/TanStack/query/blob/497804ee00927da14ee8ff7f8218f270def7a91e/packages/query-core/src/hydration.ts#L72-L74

If you do that, you can either ignore the problem I've describe above if you don't care about that, or, you can make sure that errors can be persisted as well by using something like [superjson](https://github.com/blitz-js/superjson) to serialize errors to json.

peristers themselves can define a `serialize` and `deserialize` option that just defaults to `JSON.stringify` and `JSON.parse`:

https://github.com/TanStack/query/blob/13b17ee45de3299dd3d3a2dfc3d23ad2f024882a/packages/query-sync-storage-persister/src/index.ts#L42-L43
